### PR TITLE
add support to VPC resource controller's CNINode CRD

### PR DESCRIPTION
**Pull request by haouc**
**What type of PR is this?**
We want to replace node labels when enabling SGP feature through VPC CNI. VPC resource controller will create the new CRD objects with one to one mapping to k8s nodes. VPC CNI is responsible to patch the CR to add features.
The resource controller's PR is https://github.com/aws/amazon-vpc-resource-controller-k8s/pull/270

The tests have been done with resource controller change. 

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:
We need to switch to the new CRD based implementation.

**If an issue # is not available please add repro steps and logs from IPAMD/CNI showing the issue**:

**Testing done on this change**:

- new install
- upgrade from older CNI 1.11.x and 1.12.x
- downgrade to older CNI version
- scale tests with 300 nodes
- 
**Automation added to e2e**:

Will need wait for the VPC resource controller release

**Will this PR introduce any new dependencies?**:

Yes, depending on the new CRD

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:
It will not.

**Does this change require updates to the CNI daemonset config files to work?**:

no
**Does this PR introduce any user-facing change?**:

The node's label for trunk interface will no longer be added.
```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by haouc** [comment]
**Note**: Since this is waiting for a new release tag of VPC resource controller for an update on the CNINode CRD, it is referring to my fork repo tag by `replacing`. We need remove the `replacing` before merging this PR.

File: pkg/k8sapi/k8sutils.go
Comment (jdn5126) [inline pkg/k8sapi/k8sutils.go:73]: Similar to https://github.com/aws/amazon-vpc-cni-k8s/pull/2439 , do we need to scope down caching for this CRD by adding an object selector for this node only? Alternatively, we may not want to cache this object since there will always only be one per node, and we only do patch operations

File: pkg/sgpp/utils.go
Comment (jdn5126) [inline pkg/sgpp/utils.go:70]: nit: as above, I think we can remove this function if we do not plan to call it

File: pkg/sgpp/utils.go
Comment (jdn5126) [inline pkg/sgpp/utils.go:42]: Since this package is for `sgpp`, and this feature is going to be more general, I am wondering if this belongs in a different package. Maybe `k8sapi` or `utils`?

**Review by jdn5126 — COMMENTED** [review]
Change overall LGTM, no major concerns. I added a few questions and comments. For the manifest changes, can we make the `aws-vpc-cni` chart changes as well? Since these are only additions, they are fine to add in this PR

File: pkg/ipamd/ipamd.go
Comment (ellistarn) [reply pkg/ipamd/ipamd.go:578]: Do we really want to crash the CNI? Can't we simply retry and emit a log each time? Crashing will subject us to crash loop backoff durations.

File: pkg/ipamd/ipamd.go
Comment (ellistarn) [reply pkg/ipamd/ipamd.go:584]: Another idea for logging. Consider avoiding words like "successfully" and "failed" in INFO and ERROR logs respectively. Both INFO and ERROR implicitly convey these concepts, but nature of their log level. 

e.g. `enabled feature %q for node %q`


Also -- should we be using a structured logger, so this is more easily parsable?

log.With("node", node.Name, "feature", rc.CustomNetworking).Info("enabled feature")

File: pkg/sgpp/utils.go
Comment (ellistarn) [inline pkg/sgpp/utils.go:42]: This function is not attached to any struct and is "dangling". Is it possible to attach this to a greater context, so we can avoid the complexity of passing dependencies around through method arguments (e.g. k8sClient). 

Another thought -- does it make sense to decouple these two concepts:
1. Constructs of the CNINode
2. Application of the CNINode

Something like:
```
node := v1.Node{}
_, err := client.Get(ctx, &node) { ... }
cniNode := CNINodeFor(node)
client.Patch(ctx, &cniNode) { ... }

File: pkg/sgpp/utils.go
Comment (ellistarn) [inline pkg/sgpp/utils.go:1]: As a general pattern, I'd avoid `utils` packages as much as possible, and encapsulate the logic where its owned wherever possible. Of course, some things do need to be shared, but often there is a better named abstraction that encapsulates the concept. `utils` packages have a tendency to become a dumping ground for code, and become more complex and less maintainable over time.

File: pkg/ipamd/ipamd.go
Comment (ellistarn) [reply pkg/ipamd/ipamd.go:1282]: Should this merely emit a debug log? It seems like if we need a trunk eni, but can't get one, that should be considered as an error case.

File: pkg/sgpp/utils.go
Comment (haouc) [reply pkg/sgpp/utils.go:42]: This can be added to ipamd.go instead of from another package. I prefer to create the CNINode objects from VPC resource controller side since in general VPC resource controller is persisting and tracking/managing the life time of the node object from its point of view. VPC CNI will restart for many reasons and seems no need to worry about creation of this CRD object, like it doesn't care the creation of node object.

Comment (haouc) [reply pkg/sgpp/utils.go:42]: According to Ellis' suggestion, I think it makes sense to have this function associated to IPAM.

File: pkg/sgpp/utils.go
Comment (haouc) [reply pkg/sgpp/utils.go:70]: Makes sense. I was hesitating the same and decided keeping it here to save the option to support disable the feature without terminating nodes. I can remove them.

File: pkg/k8sapi/k8sutils.go
Comment (haouc) [reply pkg/k8sapi/k8sutils.go:73]: Do you mean always use raw client for this operation?

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go:1295]: Sure, will remove it.

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go:584]: Makes sense.

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go:1286]: Yes, it is just log the error and continue. This is ok for nodeinit and the reconciling will try it again later.

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go:578]: If Get returns error of NotFound, we just log the error and move on. The reconciler will try again.

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go:587]: OK. Since we don't current plan to support remove networking without terminating the node, I will update the comment and remove the code. Thanks

File: pkg/sgpp/utils.go
Comment (ellistarn) [reply pkg/sgpp/utils.go:42]: > I prefer to create the CNINode objects from VPC resource controller side since in general VPC resource controller is persisting and tracking/managing the life time of the node object from its point of view. 

Huge +1. I could see taking this a step further, and taking an assumption that the CRD exists for the given node before taking any actions as a CNI. WDYT?

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go:1282]: I agree. I think this should be qualified for an error even it is induced by customers.

File: pkg/sgpp/utils.go
Comment (haouc) [reply pkg/sgpp/utils.go:42]: That would be nice. I am trying to find out if there is a k8s mechanism which is like a "generator" (the opposite side of finalizer) with Ownership. I think this is no such native controller to do this? In this case, we are relying on the VPC resource controller to generate the CRD object upon its owner node object's creation. Not sure if this is what you were referring to?

File: pkg/sgpp/utils.go
Comment (haouc) [reply pkg/sgpp/utils.go:1]: added algo/algo.go to utils folder. Seems ok practice referring to https://github.com/kubernetes/kubernetes/tree/master/pkg/util
Probably should use `util` instead of `utils` in VPC CNI?

File: pkg/ipamd/ipamd.go
Comment (jdn5126) [reply pkg/ipamd/ipamd.go]: In this implementation, for custom networking, it looks like we only patch the CRD during node init. So if we cannot patch during node init, it would never be patched, and if the k8s client returns an error, `aws-node` pod will crash.

It looks like only SGPP uses a reconciler approach

File: pkg/sgpp/utils.go
Comment (ellistarn) [reply pkg/sgpp/utils.go]: Yeah there's no "generator" concept. I think of the CNINode is being owned by the CNI-agent, as it is the source of truth for the CNINode.spec. I'm not sure I see the benefit of creating the object from VPCRC, regardless of the restart strategy of VPCRC or VPCCNI.

What do you think about the following:
* VPCCNI wakes up, and applies (create/update) the CNINode to be what it expects it to be.
* VPCRC watches for the CNINode objects and takes action accordingly (i.e., attaching trunk eni)

File: pkg/utils/algo/algos.go
Comment (ellistarn) [inline pkg/utils/algo/algos.go:6]: Consider using https://github.com/samber/lo#contains

File: pkg/k8sapi/k8sutils.go
Comment (ellistarn) [inline pkg/k8sapi/k8sutils.go:80]: Is this env var already defined?

File: pkg/ipamd/ipamd.go
Comment (ellistarn) [inline pkg/ipamd/ipamd.go:2318]: Check out: https://github.com/samber/lo#containsby

```
if lo.ContainsBy(cniNode.Spec.Features, func(feature rcv1alpha1.Feature) bool {
  return feature == rcv1alpha1.Feature{ Name: featureName, Value: featureValue }
}) {
  return nil
}
```

File: pkg/utils/algo/algos.go
Comment (ellistarn) [inline pkg/utils/algo/algos.go:6]: Consider using github.com/samber/lo for functional interfaces like this.

File: pkg/sgpp/utils.go
Comment (haouc) [reply pkg/sgpp/utils.go]: >the CNINode is being owned by the CNI-agent
I am not sure if this is correct since the agent will change for any reason but the CNINode should be living with the node. My preference is
- VPC RC (sort of) manage "nodes" and their owning CNINode objects
- VPC CNI doesn't have to worry about the CNINode creation since it should be taken care by VPC RC
- VPC CNI shouldn't have create permission on this resource
- VPC RC managing these objects will benefit future improvements (we can sync offline)

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go]: I don't think we support reconciling on the eniconfig label currently. Can you double check? I checked the code in master branch and seems we only do that in init.

Comment (jdn5126) [reply pkg/ipamd/ipamd.go]: Right, we do not currently support reconciling on ENIConfig label, so this is not a change in behavior, I am just pointing out that our restart strategy here is to crash the `aws-node` pod when the CRD does not exist. Which may not be the best idea as @ellistarn pointed out, but it's the least complex

File: pkg/k8sapi/k8sutils.go
Comment (haouc) [reply pkg/k8sapi/k8sutils.go:80]: It is defined in CNI config yaml.
https://github.com/aws/amazon-vpc-cni-k8s/blob/master/config/master/aws-k8s-cni.yaml#L204

File: pkg/sgpp/utils.go
Comment (ellistarn) [reply pkg/sgpp/utils.go]: Worth a discussion, I think.

File: pkg/ipamd/ipamd.go
Comment (ellistarn) [reply pkg/ipamd/ipamd.go]: Isn't it straightforward for us to write a controller-runtime reconciler that's included with the ipamd server?

Comment (jdn5126) [reply pkg/ipamd/ipamd.go]: If you mean to reconcile on the CRD, then yeah, but I think that overhead is not worth it. We have a push to decrease the amount of API server calls from IPAMD: https://github.com/aws/amazon-vpc-cni-k8s/issues/2160

Comment (ellistarn) [reply pkg/ipamd/ipamd.go]: ack, this is a good idea for daemons. As an alternative we could backoff-poll to retrieve the CRD instead of crashing the process.

Comment (jdn5126) [reply pkg/ipamd/ipamd.go]: For reference, when we refer to the "reconciliation loop", it refers to this: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/pkg/ipamd/ipamd.go#L685

This is the "main loop" for IPAMD, where we continuously check if we can reclaim IPs, need to attach new ENIs, etc.

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go:2318]: nice lib. Will use this. Thanks

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go:2312]: Sure, let me make them inline.

File: pkg/utils/algo/algos.go
Comment (haouc) [reply pkg/utils/algo/algos.go:6]: Changed to the powerful lib.

File: go.mod
Comment (ellistarn) [inline go.mod:177]: Check out go.work files, which you can add to .gitignore and not need to edit the module: https://go.dev/doc/tutorial/workspaces

Comment (haouc) [reply go.mod:177]: I will check the doc. My understanding is the multi modules workspace in this case will keep the change local and this PR will fail in GitHub actions? Which is fine since this `replace` change will be removed after the VPC RC has a new tag for CRD update. But I think I get your point.

**Review by ellistarn — COMMENTED** [review]
LGTM.

File: charts/aws-vpc-cni/Chart.yaml
Comment (jayanthvn) [inline charts/aws-vpc-cni/Chart.yaml:4]: Why is the appVersion updated? We generally update this during release..

File: charts/aws-vpc-cni/templates/clusterrole.yaml
Comment (jayanthvn) [inline charts/aws-vpc-cni/templates/clusterrole.yaml:40]: Let's discuss on adding "patch" permission offline.

File: go.mod
Comment (jayanthvn) [reply go.mod:177]: @haouc lets please make sure PR is merged after tag is released, or other option is cut a RC release on VPC RC and use that tag here. Since it will break master builds.

File: pkg/ipamd/ipamd.go
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:141]: Even this can be cleaned up right?

File: pkg/ipamd/ipamd.go
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:578]: nit -> something like `AddFeatureToCNINodeResource` or anything else maybe?

File: pkg/ipamd/ipamd.go
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:1274]: Shouldn't we add these 2 lines here -

```
podENIErrInc("askForTrunkENIIfNeeded")
log.Errorf("Failed to add SGP feature into the CNINode", err)
```

File: charts/aws-vpc-cni/templates/clusterrole.yaml
Comment (jayanthvn) [reply charts/aws-vpc-cni/templates/clusterrole.yaml:40]: Also we need to remove the `update` permission from node resource right?

File: pkg/ipamd/ipamd.go
Comment (jayanthvn) [inline pkg/ipamd/ipamd.go:2310]: I feel, we need to add some new metrics to track CNINode updates/deletes..you can refer here - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/pkg/ipamd/ipamd.go#L187

File: pkg/k8sapi/k8sutils.go
Comment (jayanthvn) [inline pkg/k8sapi/k8sutils.go:52]: Have we measured the number of API server calls increase with this change?

File: pkg/ipamd/ipamd.go
Comment (ellistarn) [reply pkg/ipamd/ipamd.go:2310]: Client go should give us these for free

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go:575]: We don't unset the features in CNINode. CNiNode itself has owner reference to the node object. When node is deleted, the CNINode will be deleted.

File: pkg/k8sapi/k8sutils.go
Comment (haouc) [reply pkg/k8sapi/k8sutils.go:52]: We are changing to use rawclient.

File: charts/aws-vpc-cni/Chart.yaml
Comment (haouc) [reply charts/aws-vpc-cni/Chart.yaml:4]: ack

File: charts/aws-vpc-cni/templates/clusterrole.yaml
Comment (haouc) [reply charts/aws-vpc-cni/templates/clusterrole.yaml:40]: We need remove node update permission. The problem is we can merge the deletion to master. I have synced with Jeff offline. We may use a feature branch instead.

**Review by jdn5126 — APPROVED** [review]

File: go.mod
Comment (haouc) [reply go.mod]: Yeah, @jayanthvn. We can discuss offline. This PR has been changed to target release branch as a possible RC release.

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go]: Cleaning work has been done in the latest update.

File: pkg/ipamd/ipamd.go
Comment (haouc) [reply pkg/ipamd/ipamd.go]: The error message has been updated in the updates.

**Pull request merged**
